cmd/download.go

```diff
@@ -333,58 +446,75 @@ func download(folder, release, url string,
+			fmt.Fprintf(os.Stdout, "Exists: %s\n", artifactTar)
+			saveToImagesFile(image.Image, image.ImageMapping, aiImages, aiImagesFile, ocpImagesFile)
+			previous++
 		} else {
-			scratchdir := path.Join(folder, "scratch")
-			_, err = os.Stat(scratchdir)
-			if err == nil {
-				err = os.RemoveAll(scratchdir)
-				if err != nil {
-					fmt.Fprintf(os.Stderr, "error: unable to remove directory %s: %e\n", path.Join(folder, image.Artifact), err)
-					os.Exit(1)
-				}
-			}
+			// Add download jobs to channel to be processed
+			jobs <- DownloadJob{image, 0}
+			totalJobs++
+		}
+	}
 
-			err = os.Mkdir(scratchdir, 0755)
+	fmt.Println()
 
-			cmd := generateSkopeoCopyCommand(scratchdir, image.Artifact, image.Image)
-			if err != nil {
-				fmt.Fprintf(os.Stderr, "error: Failed to mkdir %s: %s", scratchdir, err)
-				os.Exit(1)
-			}
+	if totalJobs == 0 {
+		fmt.Fprintf(os.Stdout, "All %d images previously downloaded.\n", totalImages)
+		os.Exit(0)
+	}
 
-			stdout, err := cmd.CombinedOutput()
-			if err != nil {
-				fmt.Fprintf(os.Stderr, "error: unable to run command %s: %s\n", strings.Join(cmd.Args, " "), string(stdout))
-				os.Exit(1)
-			}
+	if previous > 0 {
+		fmt.Fprintf(os.Stdout, "%d images previously downloaded.\n", previous)
+	}
 
-			cmd = generateTarArtifactCommand(scratchdir, image.Artifact)
-			stdout, err = cmd.CombinedOutput()
-			if err != nil {
-				fmt.Fprintf(os.Stderr, "error: unable to run command %s: %s\n", strings.Join(cmd.Args, " "), string(stdout))
-				os.Exit(1)
-			}
+	fmt.Fprintf(os.Stdout, "Queueing %d of %d images for download, with %d workers.\n\n", totalJobs, totalImages, maxParallel)
+	startTime := time.Now()
 
-			err = os.Rename(path.Join(scratchdir, artifactTar), path.Join(folder, artifactTar))
-			if err != nil {
-				fmt.Fprintf(os.Stderr, "error: unable to move file %s: %e\n", path.Join(scratchdir, artifactTar), err)
-				os.Exit(1)
-			}
+	wg := sync.WaitGroup{}
 
-			err = os.RemoveAll(scratchdir)
-			if err != nil {
-				fmt.Fprintf(os.Stderr, "error: unable to remove directory %s: %e\n", scratchdir, err)
-				os.Exit(1)
-			}
-		}
+	// Create imageDownloader workers
+	for i := 0; i < maxParallel; i++ {
+		wg.Add(1)
+		go imageDownloader(&wg, i, jobs, results, folder)
+	}
+
+	// Create results processor
+	wgResp := sync.WaitGroup{}
+	wgResp.Add(1)
+	go imageDownloaderResults(&wgResp, results, totalJobs, folder, aiImages, aiImagesFile, ocpImagesFile)
+
+	// Wait for imageDownloader workers to finish
+	wg.Wait()
 
-		saveToImagesFile(image.Image, image.ImageMapping, aiImages, aiImagesFile, ocpImagesFile)
+	// Close the jobs channel to complete the task list
+	close(jobs)
+
+	// Close results channel
+	close(results)
+	wgResp.Wait()
+
+	endTime := time.Now()
+	diff := endTime.Sub(startTime)
+
+	fmt.Fprintf(os.Stdout, "%d images of %d downloaded, with %d workers, in: %s\n", (totalJobs - downloadFailures), totalImages, maxParallel, diff.String())
```

Comment:
As commented I think we can add the images that are skipped at the end of the task. Therefore skipped + downloaded = total images. Otherwise seems weird to download less images than total.

Reply:
Added an expanded summary